### PR TITLE
Pdll

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**

### CODE REVIEW

File: tools/tpp-opt/tpp-opt.cpp [hunk 1]
```diff
@@ -30,10 +30,12 @@
 #include "TPP/Dialect/Xsmm/XsmmDialect.h"
 #include "TPP/PassBundles.h"
 #include "TPP/Passes.h"
+#include "TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmm.h"
 
 int main(int argc, char **argv) {
   mlir::registerAllPasses();
   mlir::tpp::registerTppCompilerPasses();
+  mlir::tpp::registerConvertVectorToXsmmPass();
```
Comment (adam-smnk) [inline tools/tpp-opt/tpp-opt.cpp:38]: I'd expect the pass registration to be already included in `registerTppCompilerPasses`.
Is a separate one really needed?

File: lib/TPP/Transforms/Utils/VNNIUtils.cpp [hunk 2]
```diff
@@ -36,6 +36,20 @@ bool isInVnniLayout(VnniOperandRank expectedRank, MemRefType memref) {
...
+    return false;
+  }
+  return vector.getShape().back() == vnni::utils::getVnniBlockingFactor(vector);
+}
+
+// Until we have a better way to express the VNNI layout (see: #563), it is up
+// to the callee to specify the expected rank in the VNNI layout as the rank
+// depends on the operations we are dealing with.
+bool isInVnniLayout(VnniOperandRank expectedRank, VectorType vector) {
+  return isInVnniLayout((int64_t)expectedRank, vector);
```
Comment (adam-smnk) [inline lib/TPP/Transforms/Utils/VNNIUtils.cpp:50]: nit: use static_cast

File: lib/TPP/Transforms/IntelAMXTileConfig.cpp [hunk 3]
```diff
@@ -122,17 +213,17 @@ struct IntelAMXTileConfigInsertionPass
     : public impl::IntelAMXTileConfigInsertionPassBase<
           IntelAMXTileConfigInsertionPass> {
   void populateCombinePatterns(RewritePatternSet &patterns) {
-    patterns.add<IntelAMXTileConfig<xsmm::BrgemmOp, xsmm::BrgemmDispatchOp>>(
```
Comment (adam-smnk) [inline lib/TPP/Transforms/IntelAMXTileConfig.cpp:125]: Are we ready to already retire the old lowering?

File: lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmmPDL.pdll [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+     };
+}
+
+Pattern ConvertContractToBrgemm with benefit(0),recursion{
+     let root = op<vector.contract>(input0:Op<>, input1:Op<>, input2:Op<>)->(output:TypeRange);
+     ValidateOp(root, input0, input1, input2);
+     rewrite root with{
+	let replacement = BuildOp(root, input0, input1, input2);
+        replace root with (replacement.dispatch, replacement.invoke);
+	let user = GetUser(replacement.dispatch);
```
Comment (adam-smnk) [inline lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmmPDL.pdll:104]: Not a pdll expert but AFAIK there's no validation or matching on the user here.
I don't think we can just randomly erase it when for example:
```
%0 = vector.contract
%1 = arith.subf %0, ...
```

Comment (adam-smnk) [reply lib/TPP/Conversion/ConvertVectorToXsmm/ConvertVectorToXsmmPDL.pdll:104]: I think you already match for a transfer_write as a consumer in some other patterns so, it's probably just missing here.

File: include/TPP/Dialect/Xsmm/XsmmUtils.h [hunk 5]
```diff
@@ -88,11 +120,57 @@ ArrayAttr getUnaryDispatchFlags(UnaryOp op);
...
+FailureOr<BrgemmInfo> isMappableToBrgemm(PatternRewriter &rewriter,
+                                         vector::ContractionOp contractOp,
+                                         SmallVector<Value> &inputs,
+                                         SmallVector<Value> &output,
+                                         ArrayRef<AffineMap> indexingMap);
+
+FailureOr<vector::ContractionOp>
+makeMinorDimensionsInnerMost(RewriterBase &rewriter,
+                             vector::ContractionOp contractOp, unsigned m,
+                             unsigned n, unsigned k, xsmm::DataTypeAttr type);
```
Comment (adam-smnk) [inline include/TPP/Dialect/Xsmm/XsmmUtils.h:143]: `xsmm::DataTypeAttr` needs to be removed as it still couples to Xsmm dialect

Comment (adam-smnk) [reply include/TPP/Dialect/Xsmm/XsmmUtils.h:143]: Actually, I see it still relies on `XsmmEnum` in general.
I guess for now we can keep that as it's easy to generate them and we can refactor later.

**Comment by adam-smnk** [comment]
I'm getting pretty different results between linalg vs vector to xsmm:

### Linalg test case:
`tpp-opt ../test.mlir -convert-linalg-to-xsmm -convert-xsmm-to-func | tpp-run  -e entry --entry-point-result=void -seed 123 -print`
```
func.func @entry(%arg0: memref<32x32xf32>, %arg1: memref<32x32xf32>, %arg2: memref<32x32xf32>) 
  -> memref<32x32xf32> {
  linalg.matmul ins(%arg0, %arg1: memref<32x32xf32>, memref<32x32xf32>)
    outs(%arg2: memref<32x32xf32>)
  return %arg2 : memref<32x32xf32>
}
```

### Vector test case:
`tpp-opt ../test.mlir -convert-vector-to-xsmm-pass | tpp-run  -e entry --entry-point-result=void -seed 123 -print`
```
#map = affine_map<(d0, d1, d2) -> (d0, d2)>
#map1 = affine_map<(d0, d1, d2) -> (d2, d1)>
#map2 = affine_map<(d0, d1, d2) -> (d0, d1)>
module {
  func.func @entry(%arg0: memref<32x32xf32>, %arg1: memref<32x32xf32>, %arg2: memref<32x32xf32>)
  -> memref<32x32xf32> {
    %c0 = arith.constant 0 : index
    %cst = arith.constant 0.000000e+00 : f32
    %0 = vector.transfer_read %arg0[%c0, %c0],
      %cst {in_bounds = [true, true]} : memref<32x32xf32>, vector<32x32xf32>
    %1 = vector.transfer_read %arg1[%c0, %c0],
      %cst {in_bounds = [true, true]} : memref<32x32xf32>, vector<32x32xf32>
    %2 = vector.transfer_read %arg2[%c0, %c0],
      %cst {in_bounds = [true, true]} : memref<32x32xf32>, vector<32x32xf32>
    %3 = vector.contract {indexing_maps = [#map, #map1, #map2],
      iterator_types = ["parallel", "parallel", "reduction"],
      kind = #vector.kind<add>} %0, %1, %2
      : vector<32x32xf32>, vector<32x32xf32> into vector<32x32xf32>
    vector.transfer_write %3, %arg2[%c0, %c0]
      {in_bounds = [true, true]} : vector<32x32xf32>, memref<32x32xf32>
    return %arg2 : memref<32x32xf32>
  }
}
```

**Comment by adam-smnk** [comment]
After a bit of IR diffing, the current difference comes from dispatch call:
- created from linalg:
`%0 = call @xsmm_gemm_dispatch(%c1_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c0_i64) : (i64, i64, i64, i64, i64, i64, i64, i64) -> i64`
- created from vector:
`%0 = call @xsmm_gemm_dispatch(%c1_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c1_i64, %c1_i64, %c0_i64) : (i64, i64, i64, i64, i64, i64, i64, i64, i64, i64) -> i64`

The new call builder add two extra arguments 8th and 9th `%c1_i64, %c1_i64` which correspond to unit strides.
Now I'm not really sure how this "just works"^tm as we don't have a wrapper for 10 args gemm_dispatch. But it somehow runs and just produces invalid results.

**Comment by KavithaTipturMadhu** [comment]
> After a bit of IR diffing, the current difference comes from dispatch call:
> 
> * created from linalg:
>   `%0 = call @xsmm_gemm_dispatch(%c1_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c0_i64) : (i64, i64, i64, i64, i64, i64, i64, i64) -> i64`
> * created from vector:
>   `%0 = call @xsmm_gemm_dispatch(%c1_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c32_i64, %c1_i64, %c1_i64, %c0_i64) : (i64, i64, i64, i64, i64, i64, i64, i64, i64, i64) -> i64`
> 
> The new call builder add two extra arguments 8th and 9th `%c1_i64, %c1_i64` which correspond to unit strides. Now I'm not really sure how this "just works"^tm as we don't have a wrapper for 10 args gemm_dispatch. But it somehow runs and just produces invalid results.

Fixed the issue in the last commit @adam-smnk

**Pull request closed without merging**